### DISCUSSION TITLE
using _filter_fields to push form data into database table

### DISCUSSION THREAD

**Discussion by KellerKev** [emmett-framework/emmett#324] [Q&A] [answered]
in py4web or web2py I do this to push form values into the db with the form data in a dict.

Can I do the same in emmett?

e.g. use   id = db.address.insert(**db.address._filter_fields(form.vars))

with form.vars being the form data dict. 

Thank you!

web2py sample: 

def register():
    form = SQLFORM.factory(db.client, db.address)
    if form.process().accepted:
        id = db.client.insert(**db.client._filter_fields(form.vars))
        form.vars.client = id
        id = db.address.insert(**db.address._filter_fields(form.vars))
        response.flash = 'Thanks for filling the form'
    return dict(form=form)

**Comment by gi0baro** [comment, marked as answer]
The form parameters in Emmett are contained into the `.params` attribute (https://emmett.sh/docs/2.2.x/forms)

If I got your sample right, you need to build a single form using multiple models. In Emmett I would do something like this:

```python
from emmett import Form
from emmett.helpers import flash
from emmett.orm import Model, Field, belongs_to, has_one

class Client(Model):
    has_one('address')

    first_name = Field()
    last_name = Field()

class Address(Model):
    belongs_to('client')

    city = Field()
    street = Field()

@app.route
async def register():
    fields_client = ["first_name", "last_name"]
    fields_address = ["city", "street"]
    form = await Form({
        **{
            field: Client.table[field].clone()
            for field in fields_client
        },
        **{
            field: Address.table[field].clone()
            for field in fields_address
        }
    })
    if form.accepted:
        res = Client.create(**{key: form.params[key] for key in fields_client})
        res.id.address.create(**{key: form.params[key] for key in fields_address})
        flash("thanks")
    return {"form": form}
```

**Reply by KellerKev** [reply]
Wow, thank you!! Yes, exactly I wanted to use one form + 2 tables. I did not expect it to be so different, But anyway, this is great. This is great help.

**Reply by KellerKev** [reply]
I relation to some code and the question I posted around hiding fields, I would be curious how you can do an edit form like that? 

If client_id != None (taken from the URL request param) e.h. register/1 (= editing the registration for client 1). 

Now I need to prepoulate the form with data from client 1 from the 2 tables. 
Then I need to make sure that the right columsn in the right tables get updated. 

I usually would use 2 hidden fields for that. 
edit=True or False (if edit=True I update the db, otherwise I insert which is already covered in your sample)
id= the client id shall be transferred in the post request from the form, but not shown to the user.

With the id in the form hidden, I can now update the tables.

To prepopluate the form I would usually set default values via pydal: e.g.

client_first_name.default= (Emmett orm query result dict to get first name) 

etc.

**Reply by gi0baro** [reply]
@Kkeller83 probably the answer is in the code I pushed in the other discussion https://github.com/emmett-framework/emmett/discussions/350#discussioncomment-887287
